File: mobile/modules/botcage-p2p/android/src/main/java/expo/modules/botcagep2p/BotcageP2pModule.kt
// The Android side of botcage's peer-to-peer link. A mirror of the iOS module:
// the Rust library does the work, this hands its results to JavaScript.
package expo.modules.botcagep2p

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.concurrent.thread
import uniffi.botcage_p2p.EventSink
import uniffi.botcage_p2p.Peer

class NotConnectedException :
  CodedException("this phone is not connected to a botcage yet")

class BotcageP2pModule : Module() {
  private var peer: Peer? = null

  /** Counts attempts to open the event stream, so a stale one can be ignored
   *  rather than blocking its own replacement. */
  private var generation = 0

  override fun definition() = ModuleDefinition {
    Name("BotcageP2p")
    Events("frame", "state")

    AsyncFunction("connect") { address: String ->
      val opened = Peer.connect(address)
      peer = opened
      opened.id()
    }

    AsyncFunction("request") { method: String, path: String, token: String?, body: String? ->
      val open = peer ?: throw NotConnectedException()
      val response = open.request(method, path, token, body)
      mapOf("status" to response.status.toInt(), "body" to response.body)
    }

    // Never refused because one is already running: the previous stream is
    // usually still unwinding when a retry arrives, and turning the retry away
    // silently leaves nothing to try again on.
    Function("listen") { token: String? ->
      val open = peer ?: return@Function null
      generation += 1
      val mine = generation
      open.stop()

      thread(name = "botcage.p2p.events") {
        val sink = object : EventSink {
          override fun onFrame(name: String, data: String) {
            if (mine == generation) sendEvent("frame", mapOf("name" to name, "data" to data))
          }

          override fun onState(connected: Boolean) {
            if (mine == generation) sendEvent("state", mapOf("connected" to connected))
          }
        }
        try {
          open.listen(token, sink)
        } catch (_: Exception) {
          if (mine == generation) sendEvent("state", mapOf("connected" to false))
        }
      }
      null
    }

    Function("stop") {
      peer?.stop()
      null
    }

    Function("isConnected") {
      peer != null
    }
  }
}

File: mobile/modules/botcage-p2p/ios/BotcageP2pModule.swift
// The iOS side of botcage's peer-to-peer link.
//
// Everything here is plumbing: the Rust library below does the work, and this
// hands its results to JavaScript. Requests run on Expo's background queue, and
// the event stream gets a thread of its own because it blocks until it ends.
import ExpoModulesCore

private final class Frames: EventSink {
  private weak var module: BotcageP2pModule?

  init(module: BotcageP2pModule) {
    self.module = module
  }

  func onFrame(name: String, data: String) {
    module?.emit(name: name, data: data)
  }

  func onState(connected: Bool) {
    module?.emit(connected: connected)
  }
}

public class BotcageP2pModule: Module {
  private var peer: Peer?
  private var streaming: Thread?

  func emit(name: String, data: String) {
    sendEvent("frame", ["name": name, "data": data])
  }

  func emit(connected: Bool) {
    sendEvent("state", ["connected": connected])
  }

  public func definition() -> ModuleDefinition {
    Name("BotcageP2p")
    Events("frame", "state")

    AsyncFunction("connect") { (address: String) -> String in
      do {
        let peer = try Peer.connect(address: address)
        self.peer = peer
        return peer.id()
      } catch {
        throw LinkFailed(error)
      }
    }

    AsyncFunction("request") { (method: String, path: String, token: String?, body: String?) -> [String: Any] in
      guard let peer = self.peer else {
        throw NotConnected()
      }
      do {
        let response = try peer.request(method: method, path: path, token: token, body: body)
        return ["status": Int(response.status), "body": response.body]
      } catch {
        throw LinkFailed(error)
      }
    }

    // Not async: it returns at once and the stream runs on its own thread, so
    // the JavaScript side is never left holding a promise for hours.
    Function("listen") { (token: String?) in
      guard let peer = self.peer, self.streaming == nil else { return }
      let sink = Frames(module: self)
      let thread = Thread {
        do {
          try peer.listen(token: token, sink: sink)
        } catch {
          // Whatever went wrong, the stream is not running — and the app can
          // only retry if it is told so.
          self.emit(connected: false)
        }
        self.streaming = nil
      }
      thread.name = "botcage.p2p.events"
      self.streaming = thread
      thread.start()
    }

    Function("stop") {
      self.peer?.stop()
    }

    Function("isConnected") { () -> Bool in
      self.peer != nil
    }
  }
}

private final class NotConnected: Exception {
  override var reason: String {
    "this phone is not connected to a botcage yet"
  }
}

/// Carries a Rust error's own words across to JavaScript.
///
/// Without this, Expo wraps anything it does not recognise as
/// `UnexpectedException: BotcageP2p.P2pError.Unreachable(reason: "…")` — the
/// sentence is in there, buried in a type name and a file position that mean
/// nothing to whoever is holding the phone.
private final class LinkFailed: Exception {
  private let why: String

  init(_ error: Error) {
    if let p2p = error as? P2pError {
      switch p2p {
      case let .Unreachable(reason): why = reason
      case let .Failed(reason): why = reason
      }
    } else {
      why = String(describing: error)
    }
    super.init()
  }

  required init() {
    why = "something went wrong"
    super.init()
  }

  override var reason: String { why }
}
